Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: FinalParameter

child of #14942 

I have read check documentation: https://checkstyle.org/checks/misc/finalparameters.html#FinalParameters
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> javac src/Test.java
PS D:\CS\test> cat src/Test.java
import java.util.PriorityQueue;
import java.util.Queue;

class Test {
    void test() {
        Queue<Integer> q = new PriorityQueue<>();
        q.add(1);
        for (Integer _ : q) {  // violation
            q.poll();
        }
        try {
            int x = 1 / 0;
        } catch (Exception _) {       // violation
            System.out.println("division by zero");
        }
    }
}

PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="FinalParameters">
        <property name="tokens" value="FOR_EACH_CLAUSE, LITERAL_CATCH"/>
    </module>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:8:14: Parameter _ should be final. [FinalParameters]
[ERROR] D:\CS\test\src\Test.java:13:18: Parameter _ should be final. [FinalParameters]
Audit done.
Checkstyle ends with 2 errors.


```
---

**Describe what you expect in detail.**

check doc : 

> Rationale: Changing the value of parameters during the execution of the method's algorithm can be confusing and should be avoided. A great way to let the Java compiler prevent this coding style is to declare parameters final.

JLS:

> If a declaration does not include an identifier, but instead includes the reserved keyword _ (underscore), then the entity cannot be referred to by name.

So I consider this as a false positive unnamed variables can't be referred to by name and Their value can't change so there is no point to require it to be final

---

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheck.java
@@ -48,6 +48,13 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * Default value is {@code false}.
  * </li>
  * <li>
+ * Property {@code ignoreUnnamedParameters} -
+ * Ignore <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+ * unnamed parameters</a>.
+ * Type is {@code boolean}.
+ * Default value is {@code true}.
+ * </li>
+ * <li>
  * Property {@code tokens} - tokens to check
  * Type is {@code java.lang.String[]}.
  * Validation type is {@code tokenSet}.
@@ -103,6 +110,12 @@ public class FinalParametersCheck extends AbstractCheck {
     private boolean ignorePrimitiveTypes;
 
     /**
+     * Ignore <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+     * unnamed parameters</a>.
+     */
+    private boolean ignoreUnnamedParameters = true;
+
+    /**
      * Setter to ignore primitive types as parameters.
      *
      * @param ignorePrimitiveTypes true or false.
@@ -110,6 +123,18 @@ public class FinalParametersCheck extends AbstractCheck {
      */
     public void setIgnorePrimitiveTypes(boolean ignorePrimitiveTypes) {
         this.ignorePrimitiveTypes = ignorePrimitiveTypes;
+    }
+
+    /**
+     * Setter to ignore
+     * <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+     * unnamed parameters</a>.
+     *
+     * @param ignoreUnnamedParameters true or false.
+     * @since 10.18.0
+     */
+    public void setIgnoreUnnamedParameters(boolean ignoreUnnamedParameters) {
+        this.ignoreUnnamedParameters = ignoreUnnamedParameters;
     }
 
     @Override
@@ -195,7 +220,8 @@ public class FinalParametersCheck extends AbstractCheck {
      */
     private void checkParam(final DetailAST param) {
         if (param.findFirstToken(TokenTypes.MODIFIERS).findFirstToken(TokenTypes.FINAL) == null
-                && !isIgnoredParam(param)
+                && !isIgnoredPrimitiveParam(param)
+                && !isIgnoredUnnamedParam(param)
                 && !CheckUtil.isReceiverParameter(param)) {
             final DetailAST paramName = param.findFirstToken(TokenTypes.IDENT);
             final DetailAST firstNode = CheckUtil.getFirstNode(param);
@@ -210,7 +236,7 @@ public class FinalParametersCheck extends AbstractCheck {
      * @param paramDef {@link TokenTypes#PARAMETER_DEF PARAMETER_DEF}
      * @return true if param has to be skipped.
      */
-    private boolean isIgnoredParam(DetailAST paramDef) {
+    private boolean isIgnoredPrimitiveParam(DetailAST paramDef) {
         boolean result = false;
         if (ignorePrimitiveTypes) {
             final DetailAST type = paramDef.findFirstToken(TokenTypes.TYPE);
@@ -223,6 +249,17 @@ public class FinalParametersCheck extends AbstractCheck {
             }
         }
         return result;
+    }
+
+    /**
+     *  Checks for skip current param due to <b>ignoreUnnamedParameters</b> option.
+     *
+     * @param paramDef parameter to check
+     * @return true if the parameter should be skipped due to the ignoreUnnamedParameters option.
+     */
+    private boolean isIgnoredUnnamedParam(final DetailAST paramDef) {
+        final DetailAST paramName = paramDef.findFirstToken(TokenTypes.IDENT);
+        return ignoreUnnamedParameters && paramName != null && "_".equals(paramName.getText());
     }
 
 }

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/FinalParametersCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/FinalParametersCheck.xml
@@ -19,6 +19,10 @@
             <property default-value="false" name="ignorePrimitiveTypes" type="boolean">
                <description>Ignore primitive types as parameters.</description>
             </property>
+            <property default-value="true" name="ignoreUnnamedParameters" type="boolean">
+               <description>Ignore &lt;a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html"&gt;
+ unnamed parameters&lt;/a&gt;.</description>
+            </property>
             <property default-value="METHOD_DEF,CTOR_DEF"
                       name="tokens"
                       type="java.lang.String[]"

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheckTest.java
@@ -144,4 +144,34 @@ public class FinalParametersCheckTest extends AbstractModuleTestSupport {
                 getPath("InputFinalParametersReceiver.java"), expected);
     }
 
+    @Test
+    public void testUnnamedParametersPropertyTrue() throws Exception {
+        final String[] expected = {
+            "25:18: " + getCheckMessage(MSG_KEY, "__"),
+            "30:18: " + getCheckMessage(MSG_KEY, "_e"),
+            "35:18: " + getCheckMessage(MSG_KEY, "e_"),
+            "46:14: " + getCheckMessage(MSG_KEY, "__"),
+            "49:14: " + getCheckMessage(MSG_KEY, "_i"),
+            "52:14: " + getCheckMessage(MSG_KEY, "i_"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputFinalParametersUnnamedPropertyTrue.java"), expected);
+    }
+
+    @Test
+    public void testUnnamedParametersPropertyFalse() throws Exception {
+        final String[] expected = {
+            "20:18: " + getCheckMessage(MSG_KEY, "_"),
+            "25:18: " + getCheckMessage(MSG_KEY, "__"),
+            "30:18: " + getCheckMessage(MSG_KEY, "_e"),
+            "35:18: " + getCheckMessage(MSG_KEY, "e_"),
+            "46:14: " + getCheckMessage(MSG_KEY, "__"),
+            "43:14: " + getCheckMessage(MSG_KEY, "_"),
+            "49:14: " + getCheckMessage(MSG_KEY, "_i"),
+            "52:14: " + getCheckMessage(MSG_KEY, "i_"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputFinalParametersUnnamedPropertyFalse.java"), expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/InputFinalParametersUnnamedPropertyFalse.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/InputFinalParametersUnnamedPropertyFalse.java
@@ -1,0 +1,56 @@
+/*
+FinalParameters
+ignorePrimitiveTypes = (default)false
+ignoreUnnamedParameters = false
+tokens = METHOD_DEF, CTOR_DEF, LITERAL_CATCH, FOR_EACH_CLAUSE
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.finalparameters;
+
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+public class InputFinalParametersUnnamedPropertyFalse {
+
+    void testUnnamedCatchParameter() {
+        try {
+            throw new Exception();
+        } catch (Exception _) {  // violation,'Parameter _ should be final.'
+
+        }
+        try {
+            int x = 1/0;
+        } catch (Exception __) { // violation,'Parameter __ should be final.'
+
+        }
+        try {
+            int x = 1/0;
+        } catch (Exception _e) { // violation,'Parameter _e should be final.'
+        }
+
+        try {
+            int x = 1/0;
+        } catch (Exception e_) { // violation,'Parameter e_ should be final.'
+        }
+    }
+
+    void testUnnamedForEachParameter() {
+        Queue<Integer> q = new PriorityQueue<>();
+        q.add(1);
+        q.add(2);
+        for (Integer _ : q) {  // violation,'Parameter _ should be final.'
+            var _ = q.poll();
+        }
+        for (Integer __ : q) { // violation,'Parameter __ should be final.'
+            var _ = q.poll();
+        }
+        for (Integer _i : q) { // violation,'Parameter _i should be final.'
+            var _ = q.poll();
+        }
+        for (Integer i_ : q) {  // violation,'Parameter i_ should be final.'
+           var _ = q.poll();
+        }
+    }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/InputFinalParametersUnnamedPropertyTrue.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/InputFinalParametersUnnamedPropertyTrue.java
@@ -1,0 +1,56 @@
+/*
+FinalParameters
+ignorePrimitiveTypes = (default)false
+ignoreUnnamedParameters = (default)true
+tokens = METHOD_DEF, CTOR_DEF, LITERAL_CATCH, FOR_EACH_CLAUSE
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.finalparameters;
+
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+public class InputFinalParametersUnnamedPropertyTrue {
+
+    void testUnnamedCatchParameter() {
+        try {
+            throw new Exception();
+        } catch (Exception _) {
+
+        }
+        try {
+            int x = 1/0;
+        } catch (Exception __) { // violation,'Parameter __ should be final.'
+
+        }
+        try {
+            int x = 1/0;
+        } catch (Exception _e) { // violation,'Parameter _e should be final.'
+        }
+
+        try {
+            int x = 1/0;
+        } catch (Exception e_) { // violation,'Parameter e_ should be final.'
+        }
+    }
+
+    void testUnnamedForEachParameter() {
+        Queue<Integer> q = new PriorityQueue<>();
+        q.add(1);
+        q.add(2);
+        for (Integer _ : q) {
+            var _ = q.poll();
+        }
+        for (Integer __ : q) { // violation,'Parameter __ should be final.'
+            var _ = q.poll();
+        }
+        for (Integer _i : q) { // violation,'Parameter _i should be final.'
+            var _ = q.poll();
+        }
+        for (Integer i_ : q) {  // violation,'Parameter i_ should be final.'
+           var _ = q.poll();
+        }
+    }
+}

--- a/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheckExamplesTest.java
+++ b/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheckExamplesTest.java
@@ -56,4 +56,14 @@ public class FinalParametersCheckExamplesTest extends AbstractExamplesModuleTest
         };
         verifyWithInlineConfigParser(getPath("Example3.java"), expected);
     }
+
+    @Test
+    public void testExample4() throws Exception {
+        final String[] expected = {
+            "22:12: " + getCheckMessage(MSG_KEY, "e"),
+            "44:10: " + getCheckMessage(MSG_KEY, "number"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("Example4.java"), expected);
+    }
 }

--- a/src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/Example4.java
+++ b/src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/Example4.java
@@ -1,0 +1,54 @@
+/*xml
+<module name="Checker">
+  <module name="TreeWalker">
+    <module name="FinalParameters">
+      <property name="tokens" value="FOR_EACH_CLAUSE, LITERAL_CATCH"/>
+      <property name="ignoreUnnamedParameters" value="true"/>
+    </module>
+  </module>
+</module>
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.finalparameters;
+
+// xdoc section -- start
+public class Example4 {
+
+  void testCatchParameters() {
+    try {
+      int x = 1 / 0;
+    }
+    catch (Exception e) { // violation, 'Parameter e should be final'
+      System.out.println(e);
+    }
+    try {
+      int x = 1 / 0;
+    }
+    catch (Exception _) { // ok, unnamed catch parameter, it is implicitly final
+      System.out.println("infinity");
+    }
+    try {
+      int x = 1 / 0;
+    }
+    // it is ok to have unnamed final parameters
+    // but it is unnecessary as it is implicitly final
+    catch (final Exception _) {
+      System.out.println("infinity");
+    }
+  }
+
+  void testForEachParameters() {
+    int[] l = {1, 2, 3};
+    int x = 0;
+    for (int number: l) { // violation, 'Parameter number should be final'
+      System.out.println(number);
+    }
+    // ok, unnamed enhanced for loop variable, it is implicitly final
+    for (int _: l) {
+      x++;
+    }
+  }
+
+}
+// xdoc section -- end

--- a/src/xdocs/checks/misc/finalparameters.xml
+++ b/src/xdocs/checks/misc/finalparameters.xml
@@ -42,6 +42,14 @@
               <td>6.2</td>
             </tr>
             <tr>
+              <td>ignoreUnnamedParameters</td>
+              <td>Ignore <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+ unnamed parameters</a>.</td>
+              <td><a href="../../property_types.html#boolean">boolean</a></td>
+              <td><code>true</code></td>
+              <td>10.18.0</td>
+            </tr>
+            <tr>
               <td>tokens</td>
               <td>tokens to check</td>
               <td>subset of tokens
@@ -144,6 +152,63 @@ public class Example3 {
   public void methodOne(final int x) { }
   public void methodTwo(int x) { }
   public static void main(String[] args) { } // violation, 'args should be final'
+}
+        </source>
+        <p id="Example4-config">
+          To configure the check to enforce final parameters for catch and
+          for-each blocks while ignoring unnamed parameters:
+        </p>
+        <source>
+&lt;module name=&quot;Checker&quot;&gt;
+  &lt;module name=&quot;TreeWalker&quot;&gt;
+    &lt;module name=&quot;FinalParameters&quot;&gt;
+      &lt;property name=&quot;tokens&quot; value=&quot;FOR_EACH_CLAUSE, LITERAL_CATCH&quot;/&gt;
+      &lt;property name=&quot;ignoreUnnamedParameters&quot; value=&quot;true&quot;/&gt;
+    &lt;/module&gt;
+  &lt;/module&gt;
+&lt;/module&gt;
+        </source>
+        <p id="Example4-code">
+          Example:
+        </p>
+        <source>
+public class Example4 {
+
+  void testCatchParameters() {
+    try {
+      int x = 1 / 0;
+    }
+    catch (Exception e) { // violation, 'Parameter e should be final'
+      System.out.println(e);
+    }
+    try {
+      int x = 1 / 0;
+    }
+    catch (Exception _) { // ok, unnamed catch parameter, it is implicitly final
+      System.out.println(&quot;infinity&quot;);
+    }
+    try {
+      int x = 1 / 0;
+    }
+    // it is ok to have unnamed final parameters
+    // but it is unnecessary as it is implicitly final
+    catch (final Exception _) {
+      System.out.println(&quot;infinity&quot;);
+    }
+  }
+
+  void testForEachParameters() {
+    int[] l = {1, 2, 3};
+    int x = 0;
+    for (int number: l) { // violation, 'Parameter number should be final'
+      System.out.println(number);
+    }
+    // ok, unnamed enhanced for loop variable, it is implicitly final
+    for (int _: l) {
+      x++;
+    }
+  }
+
 }
         </source>
       </subsection>

--- a/src/xdocs/checks/misc/finalparameters.xml.template
+++ b/src/xdocs/checks/misc/finalparameters.xml.template
@@ -86,6 +86,23 @@
                  value="resources/com/puppycrawl/tools/checkstyle/checks/finalparameters/Example3.java"/>
           <param name="type" value="code"/>
         </macro>
+        <p id="Example4-config">
+          To configure the check to enforce final parameters for catch and
+          for-each blocks while ignoring unnamed parameters:
+        </p>
+        <macro name="example">
+          <param name="path"
+                 value="resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/Example4.java"/>
+          <param name="type" value="config"/>
+        </macro>
+        <p id="Example4-code">
+          Example:
+        </p>
+        <macro name="example">
+          <param name="path"
+                 value="resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/finalparameters/Example4.java"/>
+          <param name="type" value="code"/>
+        </macro>
       </subsection>
 
       <subsection name="Example of Usage" id="Example_of_Usage">

